lib/puppet/util/windows/process.rb

```diff
@@ -17,7 +19,8 @@ def execute(command, arguments, stdin, stdout, stderr)
         :stdout => stdout,
         :stderr => stderr
       },
-      :close_handles => false
+      :close_handles => false,
+      :creation_flags => CREATE_NO_WINDOW,
```

Comment:
What other resource types will hardcoding this here affect?

Reply:
is there anything that should be opening a new window? I can make this configurable, but I'm not aware of a use case where a user would want to manage puppet yet pop open a window

Reply:
I think I agree, I just worry about landing a change with too large of a scope in a Z

Reply:
I can get on-board with that. What are your thoughts @Iristyle @glennsarti ?

Reply:
From docs

```
The process is a console application that is being run without a console window. Therefore, the console handle for the application is not set.
This flag is ignored if the application is not a console application, or if it is used with either CREATE_NEW_CONSOLE or DETACHED_PROCESS.
```

I'm uncertain if there are console apps that require a window. Might be worth doing a little research or tracking one down.

Generally I think the idea seems correct, but it could be a breaking change as @branan has alluded to **if** an application needs a console window allocated.

Reply:
https://stackoverflow.com/a/10753521 has some notes about the difference between `DETACH_PROCESS` and `CREATE_NO_WINDOW`

```
The difference is in what the started process can do. In both cases it won't have a console. But with the CREATE_NO_WINDOW option it can call AttachConsole(ATTACH_PARENT_PROCESS) and get access to the parent's console window (if available). That explicitly will not work when you specify DETACH_PROCESS. The only option then is for the started process to use AllocConsole() to create its own console.

Or in other words, you can be sure that the started process will never be able to chatter into your own console by using DETACH_PROCESS.

```

Reply:
@Iristyle @branan Alright, from the docs here's what I have found about this approach:

`DETACH_PROCESS` will unhook the called process from the current console, but still allow that process to call `AllocConsole` later and pop open a new console for all to see [1]

the description for `CREATE_NO_WINDOW` [2] indicates it sets the console handle to nil, so the called process won't have handles to stderr, stdout etc. but it would still be technically 'attached' to the current console and therefore `AllocConsole` wouldn't be available to it.

We can use the `STARTUP_INFO` [3] structure to push parameters to the first call to ShowWindow [4], where we can set `SW_HIDE` and "Control how the window is shown" (i.e. hide the window). However with that approach, the description for `SW_HIDE` states: "Hides the window and activates another window." I have nfc what the doc means by 'activates another window' but that seems fishy to me. 

With that, it seems like there's two possible approaches here: 

1. Set `CREATE_NO_WINDOW` and just not let the process have any handles. I don't know if there would be any issues with not having console handles, but if it helps: the mock package executable created for package tests _does_ call Console.WriteLine and despite not having a handle to the console it's not failing. 

2. Set both `DETACH_PROCESS` and then subsequently `STARTUP_INFO` and force ShowWindow to call with `SW_HIDE`. There are two things I'm worried about here: 1. the 'activates another window' bit and 2. using `SW_HIDE` with ShowWindow only ever mentions the first call to that function, rather than every time [4]. That to me means if the called process executes ShowWindow enough times it will _eventually_ open a new window.

With all that said, I _think_ that we should stick with the CREATE_NO_WINDOW approach, as it _appears_ to me simply revoking the handles to the console works and will forever.

[1] https://docs.microsoft.com/en-us/windows/console/creation-of-a-console
[2] https://docs.microsoft.com/en-us/windows/desktop/ProcThread/process-creation-flags
[3] https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/ns-processthreadsapi-_startupinfow
[4] https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-showwindow

Reply:
Additionally, I'm definitely going to make this change _only_ for the package provider, rather than all exec calls.